sootup.core/src/main/java/sootup/core/validation/InvokeArgumentValidator.java

```diff
@@ -22,31 +22,80 @@
  * #L%
  */
 
-import java.util.List;
+import sootup.core.IdentifierFactory;
+import sootup.core.jimple.basic.Immediate;
+import sootup.core.jimple.common.expr.AbstractInvokeExpr;
+import sootup.core.jimple.common.stmt.Stmt;
 import sootup.core.model.Body;
+import sootup.core.signatures.MethodSignature;
+import sootup.core.typehierarchy.TypeHierarchy;
+import sootup.core.types.ClassType;
+import sootup.core.types.PrimitiveType;
+import sootup.core.types.Type;
 import sootup.core.views.View;
 
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
 /**
  * A basic validator that checks whether the length of the invoke statement's argument list matches
  * the length of the target methods's parameter type list.
  *
  * @author Steven Arzt
  */
 public class InvokeArgumentValidator implements BodyValidator {
+    @Override
+    public List<ValidationException> validate(Body body, View view) {
+        List<ValidationException> validationException = new ArrayList<>();
+
+        for (Stmt stmt : body.getStmts()) {
+            if (stmt.containsInvokeExpr()) {
+                AbstractInvokeExpr invExpr =
+                        stmt.getInvokeExpr();
+                MethodSignature callee = invExpr.getMethodSignature();
+                List<Immediate> args = invExpr.getArgs();
+                List<Type> parameterTypes = callee.getParameterTypes();
+                if (invExpr.getArgCount() != parameterTypes.size()) {
+                    validationException.add(new ValidationException(stmt, "Invalid number of arguments"));
+                } else {
+                    // check argument type
+                    TypeHierarchy typeHierarchy = view.getTypeHierarchy();
+                    IdentifierFactory identifierFactory = view.getIdentifierFactory();
+                    Iterator<Immediate> iterArgs = args.iterator();
+                    Iterator<Type> iterParameters = parameterTypes.iterator();
+                    while (iterArgs.hasNext() && iterParameters.hasNext()) {
+                        // handle implicit conversion cases. e.g., `int` is used as an argument of a `double` parameter
+                        Type argType = iterArgs.next().getType();
+                        Type paraType = iterParameters.next();
+                        if (argType instanceof PrimitiveType && paraType instanceof PrimitiveType
+                                && PrimitiveType.isImplicitlyConvertibleTo((PrimitiveType) argType, (PrimitiveType) paraType)) {
+                            continue;
+                        }
+                        // other cases
+                        ClassType argClassType = getClassType(identifierFactory, argType);
```

Comment:
After the primitive type check the argument could still have the type: arraytype. You can not assume it is automatically a classtype.